### PR TITLE
feat(sql): Add JSON support

### LINKED ISSUE #3895: [SQL Connector] Add JSON support
__What should we do?__

We should add support for JSON types that are supported for most (if not all) databases we support with the SQL Connector. For Postgres this would be support for JSON & JSONB.

There seems to be support available via this module: https://jdbi.org/#jdbi3-json

__Why should we do it?__

We see some customer requests asking for JSON support:
https://forum.camunda.io/t/feel-expression/47520/7

### PR DESCRIPTION
**Pull request by ztefanie**
## Description
Added JSON parsing for string columns.
Added Uni Tests for all sql specific json columns

## Related issues

closes #3895 

## Checklist

- [x] PR has a **milestone** or the `no milestone` label.

### CODE REVIEW

File: connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+public class JdbiJsonHelper {
+
+  public static ResultIterable<Map<String, Object>> mapToParsedMap(Query query) {
+    return query.map(
+        (rs, ctx) -> {
+          Map<String, Object> row = new HashMap<>();
+          for (int i = 1; i <= rs.getMetaData().getColumnCount(); i++) {
+            String columnName = rs.getMetaData().getColumnLabel(i);
+            Object value = rs.getObject(i);
+            if (isPotentiallyJsonColumn(rs.getMetaData().getColumnType(i), value)) {
```
Comment (sbuettner) [inline connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java:23]: Did you try using `rs.getMetaData().getColumnTypeName`?

It might be more useful as it returns the
> type name used by the database. If the column type is a user-defined type, then a fully-qualified type name is returned.

Instead of the JDBC type returned by `getColumnType`.

Comment (ztefanie) [reply connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java:23]: Yes, @johnBgood and I have thought about this. We decided to go with the `getColumnType` function, because there we get a List of SQL types that are not depending on the SQL dialect but general for all SQL databases, which is what we want. Some SQL dialects do not have specific JSON datatypes, but use VARCHAR for JSONs. 

If we would use `getColumnTypeName` we need to check for all types of all dialects, which makes the check a lot more complex, without any additional benefit.

Comment (sbuettner) [reply connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java]: > If we would use getColumnTypeName we need to check for all types of all dialects, which makes the check a lot more complex, without any additional benefit.

But currently we try to parse simple strings as json just to figure out whether it works or not right?

File: connectors/jdbc/pom.xml [hunk 2]
```diff
@@ -38,6 +38,11 @@
       <artifactId>jdbi3-core</artifactId>
       <version>3.47.0</version>
     </dependency>
+    <dependency>
+      <groupId>org.jdbi</groupId>
+      <artifactId>jdbi3-jackson2</artifactId>
+      <version>3.47.0</version>
```
Comment (johnBgood) [inline connectors/jdbc/pom.xml:44]: we can start using a variable for this version now that we have more than 1 occurrence

File: connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java (hunk 1 above)
Comment (ztefanie) [reply connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java]: You can see in the test which dialect handels it how:

MSSQL: NVARCHAR
MYSQL: JSON
POSTGRESQL: JSON, JSONB
MARIADB: JSON

File: connectors/jdbc/src/test/java/io/camunda/connector/jdbc/integration/JdbiJdbcClientIntegrationTest.java [hunk 3]
```diff
@@ -394,4 +398,26 @@ public void shouldReturnResultList_whenSelectQueryWhereInWithBindingParameters(
       }
     }
   }
+
+  @Nested
+  class JsonTests {
+
+    @ParameterizedTest
+    @MethodSource(PROVIDE_SQL_SERVERS_CONFIG)
+    public void shouldParseJson_whenPostgresJsonAndJsonBColumnTypes(IntegrationTestConfig config)
```
Comment (johnBgood) [inline connectors/jdbc/src/test/java/io/camunda/connector/jdbc/integration/JdbiJdbcClientIntegrationTest.java:407]: I think the naming is incorrect as it's used for all dbs, right?

File: connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+              }
+            }
+            row.put(columnName, value);
+          }
+          return row;
+        });
+  }
+
+  private static boolean isPotentiallyJsonColumn(int columnType) {
+    Set<Integer> jsonCompatibleTypes =
```
Comment (johnBgood) [inline connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java:40]: this could be a constant

File: connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.jdbi.v3.core.statement.Query;
+import org.jdbi.v3.json.Json;
+
+public class JdbiJsonHelper {
+
+  public static ResultIterable<Map<String, Object>> mapToParsedMap(Query query) {
+    return query.map(
+        (rs, ctx) -> {
+          Map<String, Object> row = new HashMap<>();
+          for (int i = 1; i <= rs.getMetaData().getColumnCount(); i++) {
```
Comment (johnBgood) [inline connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java:20]: still not sure about this `1` , it might at least be worth a comment

File: connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java (hunk 1 above)
Comment (sbuettner) [reply connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java]: >  put jsons in VARCHAR columns
Yes, but thats something that needs to be handled in userland from my perspective. We should only cover when json is managed in fields we know that contain json.

Comment (sbuettner) [reply connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java]: Otherwise we are just guessing and bruteforce json deserialization on every string (plus the other types that are defined in the method).

Comment (ztefanie) [reply connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java]: ok. Thanks for explaining. I adjusted it.

File: connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+              }
+            }
+            row.put(columnName, value);
+          }
+          return row;
+        });
+  }
+
+  private static boolean isJsonColumn(String databaseProductName, String columnTypeName) {
+    Map<String, Set<String>> dbTypeMapping = new HashMap<>();
```
Comment (johnBgood) [inline connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java:42]: minor detail, but it can be a static field in this class created with Map.of(....,...,....,...)

**Review by johnBgood — DISMISSED** [review]
🥇

**Comment by johnBgood** [comment]
# QA information
## Test Environment
- Test required in (you can check both):
  - [ ] SaaS
  - [ ] SM
- [ ] Create a new test for this feature in our [Pre Release Tests](https://modeler.ultrawombat.com/projects/017779a8-9459-4feb-8b47-eeef8b2717ef--connectors-pre-release-test) for @Szik to be able to test

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*

File: connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+          // https://docs.oracle.com/en/java/javase/17/docs/api/java.sql/java/sql/ResultSetMetaData.html#getColumnTypeName(int)
+          for (int i = 1; i <= rs.getMetaData().getColumnCount(); i++) {
+            String columnName = rs.getMetaData().getColumnLabel(i);
+            Object value = rs.getObject(i);
+            if (isJsonColumn(databaseProductName, rs.getMetaData().getColumnTypeName(i))) {
+              try {
+                value =
+                    ctx.findColumnMapperFor(QualifiedType.of(JsonNode.class).with(Json.class))
+                        .orElseThrow()
+                        .map(rs, i, ctx);
```
Comment (sbuettner) [inline connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java:46]: Do we have to do this lookup on every field? It seems like its always the same mapper we are using so why not do it once for all fields and rows?

Comment (ztefanie) [reply connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java]: Good point :+1: Fixed it.

File: connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            String columnName = rs.getMetaData().getColumnLabel(i);
+            Object value = rs.getObject(i);
+            if (isJsonColumn(databaseProductName, rs.getMetaData().getColumnTypeName(i))) {
+              try {
+                value = jsonMapper.map(rs, i, ctx);
+              } catch (UnableToProduceResultException ignored) {
+                row.put(columnName, value);
+              }
+            }
+            row.put(columnName, value);
```
Comment (sbuettner) [inline connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java:52]: Shouldnt this maybe in an `else` clause?

Comment (ztefanie) [reply connectors/jdbc/src/main/java/io/camunda/connector/jdbc/model/client/JdbiJsonHelper.java:52]: No, if I move this in a else clause, the value will not be written in the row after json parsing it.

**Review by sbuettner — APPROVED** [review]
🎉
